adapters/AppLovin/AppLovinAdapter/GADMAdapterAppLovinNative.m

```diff
@@ -0,0 +1,159 @@
+#import "GADMAppLovinMediatedNativeUnifiedAd.h"
+#import "GADMAppLovinMediatedNativeAppInstallAd.h"
+
+@interface GADMAdapterAppLovinNative () <ALNativeAdLoadDelegate, ALNativeAdPrecacheDelegate>
+
+@property(nonatomic, weak) id<GADMAdNetworkConnector> connector;
+@property(nonatomic, strong) ALSdk *sdk;
+@property(nonatomic, strong) NSArray *adTypes;
+
+@end
+
+@implementation GADMAdapterAppLovinNative
+
++ (NSString *)adapterVersion {
+  return GADMAdapterAppLovinConstant.adapterVersion;
+}
+
+- (instancetype)initWithGADMAdNetworkConnector:(id<GADMAdNetworkConnector>)connector {
+  self = [super init];
+  if (self) {
+    self.connector = connector;
+    self.sdk = [GADMAdapterAppLovinUtils retrieveSDKFromCredentials:connector.credentials];
+
+    if (!self.sdk) {
+      [GADMAdapterAppLovinUtils log:@"Failed to initialize SDK"];
+    }
+  }
+  return self;
+}
+
+- (void)getBannerWithSize:(GADAdSize)adSize {
+  NSLog(@"Incorrect class called for banner request. Use GADMAdapterAppLovin for banner ad"
+        @" requeests.");
+  NSError *error = [NSError errorWithDomain:GADMAdapterAppLovinConstant.errorDomain
+                                       code:kGADErrorInvalidRequest userInfo:nil];
+  [self.connector adapter:self didFailAd:error];
+}
+
+- (void)getInterstitial {
+  NSLog(@"Incorrect class called for banner request. Use GADMAdapterAppLovin for interstitial ad"
+        @" requeests.");
+  NSError *error = [NSError errorWithDomain:GADMAdapterAppLovinConstant.errorDomain
+                                       code:kGADErrorInvalidRequest userInfo:nil];
+  [self.connector adapter:self didFailAd:error];
+}
+
+- (BOOL)isBannerAnimationOK:(GADMBannerAnimationType)animType {
+  return YES;
+}
+
++ (Class<GADAdNetworkExtras>)networkExtrasClass {
+  return [GADMAdapterAppLovinExtras class];
+}
+
+- (void)presentInterstitialFromRootViewController:(UIViewController *)rootViewController {
+}
+
+- (void)getNativeAdWithAdTypes:(NSArray *)adTypes options:(NSArray *)options {
+  if (!([adTypes containsObject:kGADAdLoaderAdTypeUnifiedNative]
+        || [adTypes containsObject:kGADAdLoaderAdTypeNativeAppInstall])) {
+    NSError *error = [NSError errorWithDomain: GADMAdapterAppLovinConstant.errorDomain
+                                         code: kGADErrorInvalidRequest userInfo: nil];
+    [self.connector adapter:self didFailAd:error];
+    return;
+  }
+
+  self.adTypes = adTypes;
+
+  [[ALSdk shared].nativeAdService loadNativeAdGroupOfCount:1 andNotify:self];
+}
+
+- (void)stopBeingDelegate {
+  self.connector = nil;
+}
+
+#pragma mark - AppLovin Native Load Delegate Methods
+
+- (void)nativeAdService:(nonnull ALNativeAdService *)service
+didFailToLoadAdsWithError
+                       :(NSInteger)code {
+  [self notifyFailureWithErrorCode:[GADMAdapterAppLovinUtils toAdMobErrorCode:(int)code]];
+}
+
+- (void)nativeAdService:(nonnull ALNativeAdService *)service didLoadAds:(nonnull NSArray *)ads {
+  if (ads.count < 1) {
```

Comment:
This is another pretty minor code style, since the value we're looking for is `0`, why not just explicitly state that? Vs providing a _range_ of numbers (anything less than 1, which isn't really possible, since it can't be negative).

Another alternative that Basil suggests is the "positive first" rule. If I recall, humans read things easier if it's a positive condition first. So:

```
if ( ads.count > 0 )
{
    [service precacheResourcesForNativeAd:[ads firstObject] andNotify:self];
}
else 
{
    [self notifyFailureWithErrorCode:kGADErrorMediationNoFill];
}
```